### PR TITLE
SI-6536 Generates super accessors X.super[Y].blah when Y is a class

### PR DESCRIPTION
**Pull request by JamesIry**
The main change here is to add another case for generating super accessors -
the case in X.super[Y].blah when X isn't the current class and Y is a class.
The change is deliberately kept as minimal as possible to reduce the chance
of breaking something in the 2.9.x line.

Additionally GenICode now detects the case when we're trying to emit
byte code that would be nonsense and warns about it. That
can safely be made an assert for 2.11.

Finally a related assert in RefChecks is beefed up to output a bit more useful
information.

review @paulp, @adriaanm

### CODE REVIEW

File: src/compiler/scala/tools/nsc/backend/icode/GenICode.scala [hunk 1]
```diff
@@ -767,11 +767,15 @@ abstract class GenICode extends SubComponent  {
         // to call super constructors explicitly and/or use their 'returned' value.
         // therefore, we can ignore this fact, and generate code that leaves nothing
         // on the stack (contrary to what the type in the AST says).
-        case Apply(fun @ Select(Super(_, mix), _), args) =>
+        case Apply(fun @ Select(Super(qual, mix), _), args) =>
+
+          if (!qual.isInstanceOf[This]) {
+            log("WARNING: super call where selector isn't 'this'. May generate invalid bytecode. Previous phases should have tansformed away this form of super.")
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/backend/icode/GenICode.scala]: Where I come from the word is "transform". NEWB

File: src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala [hunk 2]
```diff
@@ -118,7 +119,20 @@ abstract class SuperAccessors extends transform.Transform with transform.TypingT
...
+      // So, we're looking for items of the form clazz.super[mix].name (or clazz.super.name wich is seen as clazz.super[EMPTY].name.
+      // with some limitations. First, name has to be a term rather than a type.
+      // Then there are a couple of cases. If mix is empty then we only need an accessor if clazz is a trait, it's not this current class,
+      // or the validCurentOwner setting is false...which...ugh, is a mess. 
+      // If the mix is set then if it refers to a class and the clazz part isn't the current class
+      // it's not just super[mix].name then we need to generate an accessor.
+      // SI-6536 has more discussion about how this works.
+      if (name.isTermName && (mix == tpnme.EMPTY && (clazz.isTrait || clazz != currentClass || !validCurrentOwner ))
+                          || (mix != tpnme.EMPTY && !mixtpe.typeSymbol.isTrait && clazz != currentClass))
         ensureAccessor(sel)
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala]: Can you please extract a condition of this length into a local method? This has no binary compatibility implications (it's private) and is far easier to  read and maintain. I would maybe write it like this:

```
def requiresAccessor = name.isTermName && (clazz != currentClass) && (mix match {
  case tpnme.EMPTY => clazz.isTrait || !validCurrentOwner
  case _           => !mixtpe.typeSymbol.isTrait
})
```

Ha ha, "mix match".

Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala]: I borked the logic of the above, but the gist is stil sound (especially the 2-case match on mix)

File: src/compiler/scala/tools/nsc/backend/icode/GenICode.scala (hunk 1 above)
Comment (JamesIry) [inline src/compiler/scala/tools/nsc/backend/icode/GenICode.scala]: No, tansform. It's Californian.

File: src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala [hunk 3]
```diff
@@ -110,6 +110,7 @@ abstract class SuperAccessors extends transform.Transform with transform.TypingT
       val Select(sup @ Super(_, mix), name) = sel
       val sym   = sel.symbol
       val clazz = sup.symbol
+      val SuperType(_, mixtpe) = sup.tpe
```
Comment (JamesIry) [inline src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala]: Should have a better error if the assumption that sup.tpe is a SuperType turns out to be wrong.

**Comment by adriaanm** [comment]
LGTM

**Pull request merged**